src/main/host/descriptor/socket/netlink.rs

```diff
@@ -0,0 +1,1239 @@
+        args: RecvmsgArgs,
+        mem: &mut MemoryManager,
+        cb_queue: &mut CallbackQueue,
+    ) -> Result<RecvmsgReturn, SyscallError> {
+        if !args.control_ptr.ptr().is_null() {
+            log::debug!("Netlink sockets don't yet support control data for recvmsg()");
+            return Err(Errno::EINVAL.into());
+        }
+        let Some(flags) = MsgFlags::from_bits(args.flags) else {
+            warn_once_then_debug!("Unrecognized recv flags: {:#b}", args.flags);
+            return Err(Errno::EINVAL.into());
+        };
+
+        let mut packet_buffer = Vec::new();
+        let (_rv, _num_removed_from_buf) =
+            common.recvmsg(socket, &mut packet_buffer, flags, mem, cb_queue)?;
+        self.refresh_file_state(common, cb_queue);
+
+        let mut writer = IoVecWriter::new(args.iovs, mem);
+
+        // We set the source address as the netlink address of the kernel
+        let src_addr = SockaddrStorage::from_netlink(&NetlinkAddr::new(0, 0));
+
+        let buffer = (|| {
+            if packet_buffer.len() < 16 {
+                log::warn!("The processed packet is too short");
+                return self.handle_error(&packet_buffer[..]);
+            }
+            let nlmsg_type = u16::from_le_bytes((&packet_buffer[4..6]).try_into().unwrap());
+
+            match nlmsg_type {
+                RTM_GETLINK => self.handle_ifinfomsg(common, &packet_buffer[..]),
+                RTM_GETADDR => self.handle_ifaddrmsg(common, &packet_buffer[..]),
+                _ => {
+                    warn_once_then_debug!(
+                        "Found unsupported nlmsg_type: {nlmsg_type} (only RTM_GETLINK
+                        and RTM_GETADDR are supported)"
+                    );
+                    self.handle_error(&packet_buffer[..])
+                }
+            }
+        })();
+
+        // Try to write as much as we can. If the buffer is too small, just discard the rest
+        let mut total_copied = 0;
+        let mut buf = buffer.as_slice();
+        while !buf.is_empty() {
+            match writer.write(buf) {
+                Ok(0) => break,
+                Ok(n) => {
+                    buf = &buf[n..];
+                    total_copied += n;
+                }
+                Err(ref e) if e.kind() == ErrorKind::Interrupted => continue,
+                Err(e) => return Err(e.into()),
+            }
+        }
+
+        let return_val = if flags.contains(MsgFlags::MSG_TRUNC) {
+            buffer.len()
+        } else {
+            total_copied
+        };
+
+        Ok(RecvmsgReturn {
+            return_val: return_val.try_into().unwrap(),
+            addr: Some(src_addr),
+            msg_flags: 0,
+            control_len: 0,
+        })
+    }
+
+    fn handle_error(&self, bytes: &[u8]) -> Vec<u8> {
+        // If we can't get the pid, set it to zero
+        let nlmsg_seq = match bytes.get(8..12) {
+            Some(x) => u32::from_le_bytes(x.try_into().unwrap()),
```

Comment:
It's the offset of the `nlmsg_seq` field in the C struct `nlmsghdr`, see `man 7 netlink`. But that struct is not defined in Rust, so I cannot use `memoffset::span_of!`. What should I use instead? Do I need to define `nlmsghdr` in linux-api?

Reply:
> Do I need to define `nlmsghdr` in linux-api?

Ideally yes (`nlmsghdr` is in `linux/netlink.h`), but to avoid blocking this PR any longer, you can just put a comment here instead with a "TODO: " which says to replace the byte offsets with the span of the `nlmsghdr.nlmsg_seq` field, and with a perma-link to this comment (https://github.com/shadow/shadow/pull/3198#discussion_r1491978336).